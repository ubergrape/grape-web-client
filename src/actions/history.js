import findLast from 'lodash/collection/findLast'
import last from 'lodash/array/last'

import reduxEmitter from '../legacy/redux-emitter'
import * as types from '../constants/actionTypes'
import * as api from '../utils/backend/api'
import {
  userSelector, channelSelector, historySelector
} from '../selectors'
import {error} from './common'
import {showAlert, hideAlertByType} from './alert'
import * as alerts from '../constants/alerts'
import {normalizeMessage, filterEmptyMessage} from './utils'

function normalizeMessages(messages, state) {
  return messages
    .reverse()
    .map(message => normalizeMessage(message, state))
    .filter(filterEmptyMessage)
}

function loadLatest({channelId}) {
  return (dispatch, getState) => {
    // It is initial loading, show loading indicator.
    dispatch(showAlert({
      level: 'info',
      type: alerts.LOADING_HISTORY,
      delay: 1000
    }))

    api
      .loadHistory(channelId)
      .then(res => {
        const messages = normalizeMessages(res, getState())
        dispatch({
          type: types.HANDLE_INITIAL_HISTORY,
          payload: {
            messages,
            scrollTo: null
          }
        })
        dispatch(hideAlertByType(alerts.LOADING_HISTORY))
      })
      .catch(err => dispatch(error(err)))
  }
}

function loadOlder({startIndex, stopIndex, channelId}) {
  return (dispatch, getState) => {
    const {messages, olderMessages} = historySelector(getState())
    // Ensures we don't have useless requests to the backend.
    if (olderMessages) return


    const options = {
      limit: stopIndex - startIndex,
      timeTo: messages[0].time
    }

    const promise = api
      .loadHistory(channelId, options)
      .then(res => res)
      .catch(err => dispatch(error(err)))

    dispatch({
      type: types.REQUEST_OLDER_HISTORY,
      payload: promise
    })
  }
}

/**
 * This callback is called when scroller reaches the top position.
 * It renders the loaded messages when promise gets resolved.
 */
export function renderOlderHistory() {
  return (dispatch, getState) => {
    const {olderMessages} = historySelector(getState())
    olderMessages.then(res => {
      dispatch({
        type: types.HANDLE_MORE_HISTORY,
        payload: {
          messages: normalizeMessages(res, getState()),
          isScrollBack: true
        }
      })
    })
  }
}

function loadNewer({startIndex, stopIndex, channelId}) {
  return (dispatch, getState) => {
    const {messages, newerMessages} = historySelector(getState())

    // Ensures we don't have useless requests to the backend.
    if (newerMessages) return

    const options = {
      limit: stopIndex - startIndex,
      timeFrom: last(messages).time
    }

    const promise = api
      .loadHistory(channelId, options)
      .then(res => {
        dispatch({
          type: types.HANDLE_MORE_HISTORY,
          payload: {
            messages: normalizeMessages(res, getState()),
            isScrollBack: false
          }
        })
      })
      .catch(err => dispatch(error(err)))

    dispatch({
      type: types.REQUEST_NEWER_HISTORY,
      payload: promise
    })
  }
}

function loadFragment({channelId}, messageId) {
  return (dispatch, getState) => {
    const state = getState()
    const {minimumBatchSize: limit} = historySelector(state)

    api
      .loadHistoryAt(channelId, messageId, {limit})
      .then(res => {
        dispatch({
          type: types.HANDLE_INITIAL_HISTORY,
          payload: {
            messages: normalizeMessages(res, state),
            scrollTo: messageId
          }
        })
      })
      .catch(err => dispatch(error(err)))
  }
}

export function loadHistory(params) {
  return (dispatch, getState) => {
    dispatch({
      type: types.REQUEST_HISTORY,
      payload: params
    })

    if (!params.jumpToEnd) {
      if (params.startIndex !== undefined) {
        return dispatch(params.startIndex < 0 ? loadOlder(params) : loadNewer(params))
      }

      const {selectedMessageId: messageId} = historySelector(getState())
      if (messageId) return dispatch(loadFragment(params, messageId))
    }

    return dispatch(loadLatest(params))
  }
}

export function removeMessage({id: messageId}) {
  return (dispatch, getState) => {
    dispatch({
      type: types.REQUEST_REMOVE_MESSAGE,
      payload: messageId
    })
    const {id: channelId} = channelSelector(getState())
    api
      .removeMessage(channelId, messageId)
      .catch(err => dispatch(error(err)))
  }
}

export function editMessage(message) {
  return (dispatch) => {
    dispatch({
      type: types.EDIT_MESSAGE,
      payload: message
    })
    reduxEmitter.editMessage(message)
  }
}


export function editPreviousMessage() {
  return (dispatch, getState) => {
    const state = getState()
    const {messages} = historySelector(state)
    const user = userSelector(state)
    const message = findLast(messages, msg => msg.author.id === String(user.id))
    dispatch(editMessage(message))
  }
}

export function markAsUnsent(message) {
  return (dispatch) => {
    setTimeout(() => {
      dispatch({
        type: types.MARK_MESSAGE_AS_UNSENT,
        payload: message
      })
    }, 5000)
  }
}

export function createMessage({channelId, text, attachments = []}) {
  return (dispatch, getState) => {
    const state = getState()
    const id = Math.random().toString(36).substr(7)
    const author = userSelector(state)

    const message = normalizeMessage({
      id,
      text,
      author,
      time: new Date(),
      attachments,
      channel: channelId
    }, state)

    dispatch({
      type: types.ADD_PENDING_MESSAGE,
      payload: message
    })

    api
      .postMessage(channelId, text, {clientsideId: id, attachments})
      .catch(err => dispatch(error(err)))

    dispatch(markAsUnsent(message))
  }
}

export function handleMessageUpdate(message) {
  return (dispatch, getState) => {
    dispatch({
      type: types.UPDATE_MESSAGE,
      payload: normalizeMessage(message, getState())
    })
  }
}

export function resendMessage(message) {
  return (dispatch) => {
    dispatch({
      type: types.RESEND_MESSAGE,
      payload: message
    })
    dispatch(markAsUnsent(message))
  }
}

export function readMessage({id: messageId}) {
  return (dispatch, getState) => {
    dispatch({
      type: types.REQUEST_READ_MESSAGE,
      payload: messageId
    })
    const {id: channelId} = channelSelector(getState())
    api
      .readMessage(channelId, messageId)
      .catch(err => dispatch(error(err)))
  }
}
