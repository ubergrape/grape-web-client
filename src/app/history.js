import createHistory from 'history/createBrowserHistory'
import getBoundActions from './boundActions'

const history = createHistory()
history.originalPush = history.push
history.originalReplace = history.replace
history.push = (url) => {
  getBoundActions().goTo(url)
}
history.replace = (url) => {
  getBoundActions().goTo(url, {replace: true})
}

export default history
