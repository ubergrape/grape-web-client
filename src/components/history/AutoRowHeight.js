import React, {Component, PropTypes} from 'react'
import {findDOMNode, render, unmountComponentAtNode} from 'react-dom'
import bindAll from 'lodash/function/bindAll'
import noop from 'lodash/utility/noop'


/**
 * Get the height of an element.
 * If it is not rendered, use `container` to render it and detect the height.
 * If it is already rendered, detect the height in place.
 */
function getHeight(element, container, callback) {
  const node = element.render && findDOMNode(element)

  if (node) {
    callback(node.clientHeight)
    return
  }

  // Callback's context is the node.
  render(element, container, function() {
    const {clientHeight} = this
    unmountComponentAtNode(container)
    callback(clientHeight)
  })
}

const hiddenContainerStyle = {
  position: 'absolute',
  visibility: 'hidden',
  width: '100%',
  zIndex: -1
}

// Needed because container becomes relative position and needs to take max
// possible width and height.
const containerStyle = {
  width: '100%',
  height: '100%'
}

export default class AutoRowHeight extends Component {
  static propTypes = {
    cacheSize: PropTypes.number,
    rows: PropTypes.arrayOf(PropTypes.node),
    children: PropTypes.func.isRequired
  }

  static defaultProps = {
    cacheSize: 500
  }

  constructor(props) {
    super(props)
    this.updateCache([])
    this.childrenParam = bindAll(this, 'onResize', 'rowHeight', 'renderRow',
      'isRowLoaded', 'registerChild', 'recomputeRowHeights')
    this.heightsInitialized = false
  }

  componentDidMount() {
    this.hiddenContainer = findDOMNode(this.refs.hiddenContainer)
    this.calcAndCacheHeights(this.props.rows, false, () => {
      this.heightsInitialized = true
      this.forceUpdate()
    })
  }

  componentWillReceiveProps({rows}) {
    this.updateCache(rows)
    this.calcAndCacheHeights(rows)
  }

  onResize({width}) {
    // Only recalc heights when width has changed.
    if (!this.width || this.width === width) {
      this.width = width
      return
    }

    this.calcAndCacheHeights(this.props.rows, true, this.recomputeRowHeights)
  }

  registerChild(ref) {
    this.scroller = ref
  }

  recomputeRowHeights() {
    this.scroller.recomputeRowHeights()
  }

  /**
   * Get the cached height.
   */
  rowHeight(index) {
    const element = this.renderRow(index)
    return this.heights.get(element)
  }

  /**
   * A row is loaded when data is fetched and height is detected.
   */
  isRowLoaded(index) {
    const element = this.renderRow(index)
    return this.heights.has(element)
  }

  calcAndCacheHeights(rows, update, callback = noop) {
    rows.forEach((element, index) => {
      if (this.heights.has(element) && !update) {
        callback()
        return
      }

      getHeight(element, this.hiddenContainer, height => {
        this.heights.set(element, height)
        if (index === rows.length - 1) {
          callback()
        }
      })
    })
  }

  /**
   * Ensure only current elements heights are cached.
   * Otherwise we will leak refs to the old elements.
   */
  updateCache(nextElements) {
    // Key is element, value is height.
    const cache = new WeakMap()
    nextElements.forEach(element => {
      const height = this.heights.get(element)
      if (height) cache.set(element, height)
    })
    this.heights = cache
    // Iterable elements cache which is in sync with heights.
    // WeakMaps are not iterable.
    this.elements = nextElements
  }

  /**
   * Get the cached element.
   */
  renderRow(index) {
    return this.props.rows[index]
  }

  render() {
    return (
      <div style={containerStyle}>
        <div style={hiddenContainerStyle} ref="hiddenContainer"></div>
        {this.heightsInitialized && this.props.children(this.childrenParam)}
      </div>
    )
  }
}
