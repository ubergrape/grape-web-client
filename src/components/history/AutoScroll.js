import {Component, PropTypes} from 'react'
import ReactDom from 'react-dom'

/**
 * Preserves the scroll position at the bottom when messages got added.
 */
export default class AutoScroll extends Component {
  static propTypes = {
    rows: PropTypes.array.isRequired,
    bottomThreshold: PropTypes.number.isRequired,
    children: PropTypes.func.isRequired,
    rowHeight: PropTypes.func.isRequired,
    scrollTop: PropTypes.number
  }

  static defaultProps = {
    bottomThreshold: 20,
    rows: []
  }

  constructor(props) {
    super(props)
    this.childrenParam = {
      onScroll: ::this.onScroll,
      onRowsRendered: ::this.onRowsRendered,
      scrollTop: props.scrollTop
    }
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.rows !== this.props.rows) {
      // We assume user scrolled up (reverse order) and we have loaded
      // previous rows.
      // When they are inserted at the beginning, they will change our scroll
      // position, so we need to calculate the height of those rows and scroll
      // to the old position.
      const prevRenderedRow = this.props.rows[this.renderedRows.startIndex]
      const prevStartIndex = nextProps.rows.indexOf(prevRenderedRow)

      if (prevStartIndex === -1) return

      let addedHeight = 0
      for (let i = 0; i < prevStartIndex; i++) {
        const height = this.props.rowHeight(i)
        if (height) addedHeight += height
      }

      this.childrenParam.scrollTop = addedHeight
    }
  }

  onScroll({scrollHeight, clientHeight, scrollTop}) {
    this.childrenParam.scrollTop = scrollTop

    if (!clientHeight) return

    const bottomThreshold = scrollHeight - scrollTop - clientHeight
    if (bottomThreshold < this.props.bottomThreshold) {
      this.childrenParam.scrollTop = scrollTop + clientHeight
    }
  }

  onRowsRendered(params) {
    this.renderedRows = params
  }

  render() {
    return this.props.children(this.childrenParam)
  }
}
