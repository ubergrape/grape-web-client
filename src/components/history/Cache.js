import shallowEqual from 'fbjs/lib/shallowEqual'

export default class Cache {
  constructor(rows) {
    // FIXME move caching to a separate class and use some FIFO cache with limited
    // size, use cacheSize prop.
    this.cache = {}
    this.rows = rows || []
  }

  setRows(rows) {
    this.rows = rows
  }

  hasRowHeight(index) {
    const {id} = this.rows[index]
    return this.cache[id] && this.cache[id].height !== undefined
  }

  getRowHeight(index) {
    const {id} = this.rows[index]
    return this.cache[id].height
  }

  setRowHeight(index, height) {
    const {id} = this.rows[index]
    if (!this.cache[id]) this.cache[id] = {}
    this.cache[id].height = height
  }

  hasElement(index) {
    const data = this.rows[index]
    const cache = this.cache[data.id]
    const hasElement = cache && cache.element !== undefined
    return hasElement && shallowEqual(data, cache.data)
  }

  getElement(index) {
    const {id} = this.rows[index]
    return this.cache[id].element
  }

  setElement(index, element) {
    const data = this.rows[index]
    const {id} = data
    if (!this.cache[id]) this.cache[id] = {}
    this.cache[id].element = element
    this.cache[id].data = data
  }
}
