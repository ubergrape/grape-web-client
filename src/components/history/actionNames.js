export default {
  loadHistory: 'onLoad',
  renderOlderHistory: 'onTouchTopEdge',
  unsetHistoryScrollTo: 'onUserScrollAfterScrollTo',
  removeMessage: 'onRemove',
  editMessage: 'onEdit',
  resendMessage: 'onResend',
  readMessage: 'onRead'
}
