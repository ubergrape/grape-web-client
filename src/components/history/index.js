export {styles as baseMessageTheme} from './messages/baseMessageTheme'
export {OwnBubble} from './messages/regular/Bubbles'
