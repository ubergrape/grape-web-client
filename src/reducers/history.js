import * as types from '../constants/actionTypes'
import reject from 'lodash/collection/reject'
import findIndex from 'lodash/array/findIndex'
import uniq from 'lodash/array/uniq'

const initialState = {
  messages: [],
  minimumBatchSize: 30
}

function updateMessage(state, newMessage) {
  const {messages} = state
  const index = findIndex(messages, {id: newMessage.id})
  const currMessage = messages[index]
  if (index === -1) return state
  const message = {...currMessage, ...newMessage}
  messages.splice(index, 1, message)
  return {...state, messages: [...messages]}
}

/**
 * Mark last message as read, set `state` on all other messages to `undefined`
 * to remove the status mark.
 */
function markLastMessageAsRead(messages, senderId) {
  return messages.map((message, index) => {
    if (!message.state || message.author.id === senderId) return message
    return {
      ...message,
      state: index === messages.length - 1 ? 'read' : undefined
    }
  })
}

export default function reduce(state = initialState, action) {
  const {payload} = action
  switch (action.type) {
    case types.SET_USER:
      return {...state, user: payload}
    case types.GO_TO_CHANNEL:
      // Clicked on the current channel.
      if (payload === state.channel.slug) return state
      return {...state, messages: []}
    case types.SET_CHANNEL:
      return {
        ...state,
        channel: payload.channel,
        selectedMessageId: payload.messageId,
        olderMessages: undefined,
        newerMessages: undefined,
        noContent: false
      }
    case types.HANDLE_INITIAL_HISTORY:
      return {
        ...state,
        ...payload,
        noContent: payload.messages.length === 0
      }
    case types.HANDLE_MORE_HISTORY: {
      const {messages: newMessages, isScrollBack} = payload
      if (!newMessages.length) return state

      let messages
      let {olderMessages, newerMessages} = state

      if (isScrollBack) {
        messages = [...newMessages, ...state.messages]
        olderMessages = undefined
      } else {
        messages = [...state.messages, ...newMessages]
        newerMessages = undefined
      }

      messages = uniq(messages, 'id')

      return {
        ...state,
        messages,
        scrollTo: null,
        olderMessages,
        newerMessages,
        noContent: false
      }
    }
    case types.REQUEST_LATEST_HISTORY:
      return {...state, messages: []}
    case types.REQUEST_OLDER_HISTORY:
      return {...state, olderMessages: payload.promise}
    case types.REQUEST_NEWER_HISTORY:
      return {...state, newerMessages: payload.promise}
    case types.UNSET_HISTORY_SCROLL_TO:
      return {...state, scrollTo: null}
    case types.REMOVE_MESSAGE:
      return {...state, messages: reject(state.messages, {id: payload})}
    case types.EDIT_MESSAGE:
      return updateMessage(state, {...payload, isSelected: true})
    case types.UPDATE_MESSAGE:
      return updateMessage(state, payload)
    case types.MARK_MESSAGE_AS_UNSENT:
      return updateMessage(state, {...payload, state: 'unsent'})
    case types.RESEND_MESSAGE:
      return updateMessage(state, {
        ...payload,
        state: 'pending'
      })
    case types.MARK_MESSAGE_AS_SENT:
      return updateMessage(state, {id: payload.messageId, state: 'sent'})
    case types.MARK_CHANNEL_AS_READ: {
      // Currently backend logic is designed to mark all messages as read once
      // user read something in that channel.
      // This is not very accurate and might change in the future. So lets not couple
      // the rest of the logic with this design and use data structure which allows
      // individual messages to have different states.
      const {channelId, isCurrentUser, userId} = payload
      if (channelId !== state.channel.id || isCurrentUser) {
        return state
      }

      return {
        ...state,
        messages: markLastMessageAsRead(state.messages, userId)
      }
    }
    case types.HANDLE_OUTGOING_MESSAGE:
      return {
        ...state,
        messages: [
          ...state.messages,
          {...payload, state: 'pending'}
        ],
        noContent: false
      }
    case types.ADD_NEW_MESSAGE: {
      if (payload.channelId !== state.channel.id) return state
      const scrollTo = payload.author.id === state.user.id ? payload.id : null
      return {
        ...state,
        scrollTo,
        messages: [...state.messages, payload],
        noContent: false
      }
    }
    default:
      return state
  }
}
