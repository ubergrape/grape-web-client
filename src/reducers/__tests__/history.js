import history from '../history'
import {
  ADD_NEW_MESSAGE,
  CLEAR_HISTORY,
  REQUEST_NEWER_HISTORY,
  REQUEST_OLDER_HISTORY,
  UNSET_HISTORY_SCROLL_TO,
  UPDATE_MESSAGE,
  HANDLE_MORE_HISTORY,
} from '../../constants/actionTypes'

describe('history reducer', () => {
  it('should handle UNSET_HISTORY_SCROLL_TO', () => {
    expect(
      history(
        { scrollToAlignment: 'end', scrollTo: 'abc' },
        { type: UNSET_HISTORY_SCROLL_TO },
      ),
    ).toMatchSnapshot()
  })

  describe('ADD_NEW_MESSAGE', () => {
    it('should not update the state on a different channel', () => {
      expect(
        history(
          {},
          {
            type: ADD_NEW_MESSAGE,
            payload: {
              message: {
                channelId: 'bbb',
                id: 'bbb',
              },
              currentChannelId: 'aaa',
            },
          },
        ),
      ).toMatchSnapshot()
    })

    it('should set scrollTo if the author is the current user', () => {
      expect(
        history(
          {
            messages: [],
          },
          {
            type: ADD_NEW_MESSAGE,
            payload: {
              message: {
                id: 'msgId1',
                author: { id: 'userId1' },
                channelId: 'channelId1',
              },
              currentUserId: 'userId1',
              currentChannelId: 'channelId1',
            },
          },
        ),
      ).toMatchSnapshot()
    })

    it('should not set scrollTo if the author is a different user', () => {
      expect(
        history(
          {
            messages: [],
          },
          {
            type: ADD_NEW_MESSAGE,
            payload: {
              message: {
                id: 'msgId1',
                author: { id: 'userId2' },
                channelId: 'channelId1',
              },
              currentUserId: 'userId1',
              currentChannelId: 'channelId1',
            },
          },
        ),
      ).toMatchSnapshot()
    })

    it('should update the message based on the clientsideId and set the state of the message to sent', () => {
      expect(
        history(
          {
            messages: [
              { id: '234' },
              { clientsideId: 'abc' },
              { clientsideId: 'def' },
            ],
          },
          {
            type: ADD_NEW_MESSAGE,
            payload: {
              message: {
                id: '123',
                author: { id: 'userId2' },
                clientsideId: 'abc',
                channelId: 'channelId1',
              },
              currentUserId: 'userId1',
              currentChannelId: 'channelId1',
            },
          },
        ),
      ).toMatchSnapshot()
    })

    it('should add the new message if the clientsideId can not be found', () => {
      expect(
        history(
          {
            messages: [
              { id: '234' },
              { id: '345', clientsideId: 'abc' },
              { id: '456', clientsideId: 'def' },
            ],
          },
          {
            type: ADD_NEW_MESSAGE,
            payload: {
              message: {
                id: '123',
                author: { id: 'userId2' },
                clientsideId: 'xyz',
                channelId: 'channelId1',
              },
              currentUserId: 'userId1',
              currentChannelId: 'channelId1',
            },
          },
        ),
      ).toMatchSnapshot()
    })

    it('should add the new message if clientsideId is not available', () => {
      expect(
        history(
          {
            messages: [{ clientsideId: 'abc' }],
          },
          {
            type: ADD_NEW_MESSAGE,
            payload: {
              message: {
                id: '123',
                channelId: 'channelId1',
                author: { id: 'userId2' },
              },
              currentUserId: 'userId1',
              currentChannelId: 'channelId1',
            },
          },
        ),
      ).toMatchSnapshot()
    })
  })

  describe('HANDLE_MORE_HISTORY', () => {
    it('should remove duplicates for messages with same id', () => {
      expect(
        history(
          {
            messages: [{ id: '123' }, { id: '234' }, { id: '345' }],
          },
          {
            type: HANDLE_MORE_HISTORY,
            payload: { messages: [{ id: '123' }, { id: '456' }] },
          },
        ),
      ).toMatchSnapshot()
    })
  })

  describe('UPDATE_MESSAGE', () => {
    it('should update the message', () => {
      expect(
        history(
          {
            messages: [
              { id: '123', text: 'Hello!' },
              { id: '234', text: 'Hello!!' },
              { id: '345', text: 'Hello!!!' },
            ],
          },
          {
            type: UPDATE_MESSAGE,
            payload: { id: '234', text: 'Hello World!' },
          },
        ),
      ).toMatchSnapshot()
    })

    it('should return the original state if the message is missing', () => {
      expect(
        history(
          {
            messages: [
              { id: '123', text: 'Hello!' },
              { id: '234', text: 'Hello!!' },
              { id: '345', text: 'Hello!!!' },
            ],
          },
          {
            type: UPDATE_MESSAGE,
            payload: { id: '999', text: 'Hello World!' },
          },
        ),
      ).toMatchSnapshot()
    })
  })

  describe('CLEAR_HISTORY', () => {
    it('should reset messages and set loadedNewerMessage to false', () => {
      expect(history({}, { type: CLEAR_HISTORY })).toMatchSnapshot()
    })
  })

  describe('REQUEST_OLDER_HISTORY', () => {
    it('should add the payload promise to olderMessagesRequest and set loadedNewerMessage to false', () => {
      expect(
        history(
          {},
          {
            type: REQUEST_OLDER_HISTORY,
            payload: {
              promise: Promise.resolve({}),
            },
          },
        ),
      ).toMatchSnapshot()
    })
  })

  describe('REQUEST_NEWER_HISTORY', () => {
    it('should add the payload promise to newerMessagesRequest and set loadedNewerMessage to false', () => {
      expect(
        history(
          {},
          {
            type: REQUEST_NEWER_HISTORY,
            payload: {
              promise: Promise.resolve({}),
            },
          },
        ),
      ).toMatchSnapshot()
    })
  })
})
