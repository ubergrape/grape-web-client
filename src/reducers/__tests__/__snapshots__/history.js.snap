// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`history reducer ADD_NEW_MESSAGE should add the new message if clientsideId is not available 1`] = `
Object {
  "channel": Object {
    "id": "channelId1",
  },
  "loadedNewerMessage": false,
  "messages": Array [
    Object {
      "clientsideId": "abc",
    },
    Object {
      "author": Object {
        "id": "userId2",
      },
      "channelId": "channelId1",
      "id": "123",
    },
  ],
  "scrollTo": null,
  "scrollToAlignment": null,
  "showNoContent": false,
  "user": Object {
    "id": "userId1",
  },
}
`;

exports[`history reducer ADD_NEW_MESSAGE should add the new message if the clientsideId can not be found 1`] = `
Object {
  "channel": Object {
    "id": "channelId1",
  },
  "loadedNewerMessage": false,
  "messages": Array [
    Object {
      "id": "234",
    },
    Object {
      "clientsideId": "abc",
    },
    Object {
      "clientsideId": "def",
    },
    Object {
      "author": Object {
        "id": "userId2",
      },
      "channelId": "channelId1",
      "clientsideId": "xyz",
      "id": "123",
    },
  ],
  "scrollTo": null,
  "scrollToAlignment": null,
  "showNoContent": false,
  "user": Object {
    "id": "userId1",
  },
}
`;

exports[`history reducer ADD_NEW_MESSAGE should not set scrollTo if the author is a different user 1`] = `
Object {
  "channel": Object {
    "id": "channelId1",
  },
  "loadedNewerMessage": false,
  "messages": Array [
    Object {
      "author": Object {
        "id": "userId2",
      },
      "channelId": "channelId1",
      "id": "msgId1",
    },
  ],
  "scrollTo": null,
  "scrollToAlignment": null,
  "showNoContent": false,
  "user": Object {
    "id": "userId1",
  },
}
`;

exports[`history reducer ADD_NEW_MESSAGE should not update the state on a different channel 1`] = `
Object {
  "channel": Object {
    "id": "aaa",
  },
}
`;

exports[`history reducer ADD_NEW_MESSAGE should set scrollTo if the author is the current user 1`] = `
Object {
  "channel": Object {
    "id": "channelId1",
  },
  "loadedNewerMessage": false,
  "messages": Array [
    Object {
      "author": Object {
        "id": "userId1",
      },
      "channelId": "channelId1",
      "id": "msgId1",
    },
  ],
  "scrollTo": "msgId1",
  "scrollToAlignment": null,
  "showNoContent": false,
  "user": Object {
    "id": "userId1",
  },
}
`;

exports[`history reducer ADD_NEW_MESSAGE should update the message based on the clientsideId and set the state of the message to sent 1`] = `
Object {
  "channel": Object {
    "id": "channelId1",
  },
  "loadedNewerMessage": false,
  "messages": Array [
    Object {
      "id": "234",
    },
    Object {
      "author": Object {
        "id": "userId2",
      },
      "channelId": "channelId1",
      "clientsideId": "abc",
      "id": "123",
      "state": "sent",
    },
    Object {
      "clientsideId": "def",
    },
  ],
  "scrollTo": null,
  "scrollToAlignment": null,
  "showNoContent": false,
  "user": Object {
    "id": "userId1",
  },
}
`;

exports[`history reducer CLEAR_HISTORY should reset messages and set loadedNewerMessage to false 1`] = `
Object {
  "loadedNewerMessage": false,
  "messages": Array [],
}
`;

exports[`history reducer REQUEST_NEWER_HISTORY should add the payload promise to newerMessagesRequest and set loadedNewerMessage to false 1`] = `
Object {
  "loadedNewerMessage": false,
  "newerMessagesRequest": Promise {},
}
`;

exports[`history reducer REQUEST_OLDER_HISTORY should add the payload promise to olderMessagesRequest and set loadedNewerMessage to false 1`] = `
Object {
  "loadedNewerMessage": false,
  "olderMessagesRequest": Promise {},
}
`;

exports[`history reducer UPDATE_MESSAGE should return the original state if the message is missing 1`] = `
Object {
  "messages": Array [
    Object {
      "id": "123",
      "text": "Hello!",
    },
    Object {
      "id": "234",
      "text": "Hello!!",
    },
    Object {
      "id": "345",
      "text": "Hello!!!",
    },
  ],
}
`;

exports[`history reducer UPDATE_MESSAGE should update the message 1`] = `
Object {
  "loadedNewerMessage": false,
  "messages": Array [
    Object {
      "id": "123",
      "text": "Hello!",
    },
    Object {
      "id": "234",
      "text": "Hello World!",
    },
    Object {
      "id": "345",
      "text": "Hello!!!",
    },
  ],
}
`;

exports[`history reducer should handle UNSET_HISTORY_SCROLL_TO 1`] = `
Object {
  "loadedNewerMessage": false,
  "scrollTo": null,
  "scrollToAlignment": null,
}
`;
